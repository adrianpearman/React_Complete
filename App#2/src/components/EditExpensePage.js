import React from 'react'

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      <h1>{props.match.params.id}</h1>
    </div>
)}

export default EditExpensePage
