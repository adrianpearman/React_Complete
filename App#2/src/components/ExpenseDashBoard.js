import React from 'react'

const ExpenseDashboard = () => (
  <div>
    <h1>homepage</h1>
  </div>
)

export default ExpenseDashboard
