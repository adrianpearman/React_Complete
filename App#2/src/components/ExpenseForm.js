import React, {Component} from 'react'
import moment from 'moment'
import { DateRangePicker, SingleDatePicker, DayPickerRangeController } from 'react-dates'

// const date = new Date()
const date = moment()
// console.log(date.format('MMM Do YYYY'));


class ExpenseForm extends Component{
  constructor(props){
    super(props)

    this.state = {
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : ' ',
        amount: props.expense ? (props.expense.amount / 100).toString() : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment() ,
        calendarFocused: false,
        submitError: ''
      }
  }


  // react-dates funtion to allow for dates to be selected before current date
  isOutsideRange = () => {
    return false
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description: description }))
  }

  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note: note}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount: amount }))
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt: createdAt }))
    }
  }

  onFocusChange = ({focused}) => {
    this.setState(() => ({ calendarFocused: focused }))
  }

  onExpenseSubmit = (e) => {
    e.preventDefault()
    // insures that the expenses have the right information before submitting
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ submitError: 'Please provide a description and amount'}))
    } else {
      this.setState(() => ({ submitError: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10)*100, //changes the value from a string a float value
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render(){
    return (
        <form className='form' onSubmit={this.onExpenseSubmit}>
          { this.state.submitError ? <p className='form__error'>{this.state.submitError}</p> : null }
          <input
            className='text_input'
            type='text'
            placeholder='Description'
            value={this.state.description}
            autoFocus
            onChange={this.onDescriptionChange}/>
          <input
            className='text_input'
            type='number'
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange = {this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={this.isOutsideRange}
          />
          <textarea
            className='textarea'
            placeholder='Leave a note for your expense (Optional)'
            onChange={this.onNoteChange}
          ></textarea>
          <div>
            <button className='button'>{this.props.buttonText}</button>
          </div>
        </form>
    )
  }
}

export default ExpenseForm
