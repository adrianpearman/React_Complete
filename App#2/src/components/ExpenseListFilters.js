import React, { Component } from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../redux/actions/filters'
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  // onFocusChange = (calendarFocused) => {
  //   this.setState(() => ({ calendarFocused }));
  // }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused: calendarFocused }))
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }

  render(){
    return(
      <div className='content-container'>
        <div className='input-group'>
          <div className='input-group__item'>
            <input
              className='text_input'
              type='text'
              placeholder='Search expenses'
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className='input-group__item'>
            <select
              className='select'
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value='date'>Date</option>
              <option value='amount'>Amount</option>
            </select>
          </div>
          <div className='input-group__item'>
            <DateRangePicker
              startDate = {this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              showClearDates={true}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}

const mapDispathToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispathToProps)(ExpenseListFilters)
