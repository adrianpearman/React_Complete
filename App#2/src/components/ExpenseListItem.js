import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import { Link } from 'react-router-dom'

const ExpenseListItem = (props) => {
  return (
    <Link to={`/edit/${props.id}`} className='list-item'>
      <div>
        <h3 className='list-item__title'>Description: {props.description}</h3>
        <span className='list-item__subtitle'>
            <p>Date: {moment(props.createdAt).format('MMMM Do, YYYY')}</p>
        </span>
      </div>
      <h3 className='list-item__data'>Amount: {numeral(props.amount / 100).format('$0,0.00')}</h3>
    </Link>
  )
}

export default ExpenseListItem
