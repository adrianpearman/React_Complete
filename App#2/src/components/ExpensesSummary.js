import React from 'react'
import numeral from 'numeral'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import selectExpenses from '../redux/selectors/expenses'
import selectExpensesTotal from '../redux/selectors/expenses-total'

export const ExpensesSummary = (props) => {
  const expenseWord = props.expensesCount === 1 ? 'expense' : 'expenses'
  const formattedExpensesTotal = numeral(props.expensesTotal/100).format('$0,0.00')

  return(
    <div className='pageHeader'>
      <div className='content-container'>
        <h1 className='pageHeader__title'>
          Viewing <span>{props.expensesCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span>
        </h1>
        <div className='pageHeader__actions'>
          <Link to='/create' className='button'>
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
