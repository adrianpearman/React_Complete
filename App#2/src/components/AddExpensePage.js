import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../redux/actions/expenses'

export class AddExpensePage extends Component{
  startAddExpense = (expense) => {
    // uses the redux store to move the data to the store
    // props.dispatch(addExpense(expense))
    this.props.startAddExpense(expense) // changed to be able to test the component. same functionality.
    // moves the user the home page after the form submitting
    this.props.history.push('/')
  }
  render(){
    return (
      <div>
        <div className='pageHeader'>
          <div className='content-container'>
            <h1 className='pageHeader__title'>Add Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            onSubmit={this.startAddExpense}
            buttonText='Add Expense'
          />
        </div>
      </div>
    )
  }
}

const mapDispathToProps = (dispatch) => {
  return{
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  }
}

export default connect(undefined, mapDispathToProps)(AddExpensePage)
