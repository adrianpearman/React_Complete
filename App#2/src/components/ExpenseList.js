import React from 'react'
import { connect } from 'react-redux' // used to connect to the redux store

import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../redux/selectors/expenses'

export const ExpenseList = (props) => {
  let Expenses = props.expenses.map((expense, index) => {
      return(
        <ExpenseListItem
          key={index}
          id={expense.id}
          description={expense.description}
          createdAt={expense.createdAt}
          amount={expense.amount}
        />
      )
  })

  return (
    <div className='content-container'>
      <div className='list-header'>
        <div className='show-for-mobile'>Expenses</div>
        <div className='show-for-desktop'>Expense</div>
        <div className='show-for-desktop'>Amount</div>
      </div>

      <div className='list-body'>
        { props.expenses.length === 0 ? (
          <div className='list-item--message'>
            <span> Currently there are no Expenses </span>
          </div>
        ) :
        <div>
          {Expenses} 
        </div>
      }
      </div>

    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList)
