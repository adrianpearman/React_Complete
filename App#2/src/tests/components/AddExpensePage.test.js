import React from 'react'
import { shallow } from 'enzyme'
import { AddExpensePage } from '../../components/AddExpensePage'
import expenses from '../fixtures/expenses'

let startAddExpense, history, wrapper;

// refactored the variables in the test suite
beforeEach(() => {
  startAddExpense = jest.fn()
  history = { push: jest.fn()}
  wrapper =  shallow(<AddExpensePage startAddExpense={startAddExpense} history={history} />)
})

test('should render Add Expense Page', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should render Add Expense Page', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1])
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(startAddExpense).toHaveBeenLastCalledWith(expenses[1])
})
