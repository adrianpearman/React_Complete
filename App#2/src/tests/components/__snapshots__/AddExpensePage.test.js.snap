// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Add Expense Page 1`] = `
<div>
  <div
    className="pageHeader"
  >
    <div
      className="content-container"
    >
      <h1
        className="pageHeader__title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      buttonText="Add Expense"
      onSubmit={[Function]}
    />
  </div>
</div>
`;
